compiler-core/src/javascript/typescript.rs

```diff
@@ -0,0 +1,530 @@
+
+            Type::App {
+                name, args, module, ..
+            } => self.print_type_app(name, args, module),
+
+            Type::Fn { args, retrn } => self.print_fn(args, retrn),
+
+            Type::Tuple { elems } => tuple(elems.iter().map(|e| self.print(e))),
+        }
+    }
+
+    fn print_var(&self, type_: &TypeVar) -> Document<'static> {
+        match type_ {
+            TypeVar::Generic { .. } | TypeVar::Unbound { .. } => "any".to_doc(),
+            TypeVar::Link { type_: typ } => self.print(typ),
+        }
+    }
+
+    /// Prints a type coming from the Gleam prelude module. These are often the
+    /// low level types the rest of the type system are built up from. If there
+    /// is no built-in Typescript equivalent, the type is prefixed with "$gleam."
+    /// and the Gleam prelude namespace will be imported during the code emission.
+    ///
+    fn print_prelude_type(&self, name: &str, args: &[Arc<Type>]) -> Document<'static> {
+        match name {
+            "Nil" => "null".to_doc(),
+            "Int" | "Float" => "number".to_doc(),
+            "UtfCodepoint" => {
+                self.tracker.borrow_mut().prelude_used = true;
+                "$gleam.UtfCodepoint".to_doc()
+            }
+            "String" => "string".to_doc(),
+            "Bool" => "boolean".to_doc(),
+            "BitString" => {
+                self.tracker.borrow_mut().prelude_used = true;
+                "$gleam.BitString".to_doc()
+            }
+            "List" => {
+                self.tracker.borrow_mut().prelude_used = true;
+                let arg0 = self.print(args.get(0).expect("print_prelude_type list"));
+                docvec!["$gleam.List<", arg0, ">"]
+            }
+            "Result" => {
+                self.tracker.borrow_mut().prelude_used = true;
+                let arg_ok = self.print(args.get(0).expect("print_prelude_type result ok"));
+                let arg_err = self.print(args.get(1).expect("print_prelude_type result err"));
+                docvec!["$gleam.Result<", arg_ok, ", ", arg_err, ">"]
+            }
+            // Getting here sholud mean we either forgot a built-in type or there is a
+            // compiler error
+            name => panic!("{} is not a built-in type.", name),
+        }
+    }
+
+    /// Prints a "named" programmer-defined Gleam type into the Typescript
+    /// equivalent.
+    ///
+    fn print_type_app(
+        &self,
+        name: &str,
+        args: &[Arc<Type>],
+        module: &[String],
+    ) -> Document<'static> {
+        let name = format!("{}$", ts_safe_type_name(name.to_camel_case()));
+        let name = match module == self.current_module {
+            true => Document::String(name),
+            false => {
+                // If type comes from a seperate module, use that module's nam
+                // as a Typescript namespace prefix
+                let module_name = module
+                    .last()
```

Comment:
What happens if two modules have the same final segment in their name?

Reply:
Yeah, I wondered about that as well. Right now it would cause issues. What I was thinking is respecting the `as_name` field during the import statement, but here when looking at the `App` type_ I only have access to the original module vector rather than its alias. So I think when I collect the import I would need to build up some hash map of `&[String] => String`. But can use you a pointer as a key like that in Rust?

Reply:
I think so, but we could also just convert it to a string if that's a problem.